tools/explorer/lib/ontology.py

```diff
@@ -31,35 +32,137 @@ def GetFieldsForTypeName(self,
+          entity_type_fields.append(field)
+
+    return entity_type_fields
+
+  #TODO: Expand NONE matching logic
+  def _CreateMatch(
+      self,
+      field_list: List[StandardField],
+      entity_type: EntityType
+  ) -> str:
+    """Determines the closeness of the match between
+    the field list and the entity type
 
     Args:
-       namespace: the name of the namespace as a string.
-       entity_type_name: the name of the entity type as a string.
-       required_only: when true will return only required fields for a given
-         type.
+      field_list: a list of StandardField objects
+      entity_type: An EntityType object
 
     Returns:
-            result_fields: a list of StandardField tuples
+      The match type as a string
     """
-    pass
+    match_type = ''
+
+    entity_field_optwrappers = entity_type.GetAllFields().values()
+    entity_field_set = {
+        StandardField(entity_type.namespace.namespace, ow.field.field)
+        for ow in entity_field_optwrappers
+    }
+    input_field_set = set(field_list)
+
+    if entity_field_set.issubset(input_field_set):
+      if input_field_set.issubset(entity_field_set):
+        match_type = 'EXACT'
+    elif entity_field_set.issubset(input_field_set):
+      if not input_field_set.issubset(entity_field_set):
+        match_type = 'CLOSE'
+    elif not entity_field_set.issubset(input_field_set):
+      if input_field_set.issubset(entity_field_set):
+        match_type = 'INCOMPLETE'
+    else:
+      match_type = 'NONE'
+    my_match = Match(field_list, entity_type, match_type=match_type)
+    return my_match
 
   def GetEntityTypesFromFields(self,
                                field_list: List[StandardField],
-                               general_type: str = None) -> List[EntityType]:
-    """Get a list of EntityType objects matching a list of StandardField tuples.
+                               general_type: str = None,
+                               match_type: str = None) -> List[EntityType]:
+    """Gets a list of EntityType objects matching
+       a given a list of StandardField tuples
 
-    Args:
-        field_list: a list of StandardField tuples to match to an entity
-        general_type: a string indicating a general type name to narrow return
-          results
+       Args:
+            field_list: a list of StandardField tuples to match to an entity
+            general_type: a string indicating a general type name
+            to narrow return results
 
-    Returns:
-          entities: a list of EntityType objects
-          matching the provided list of fields
+       Returns:
+            entities: a list of EntityType objects
+            matching the provided list of fields
     """
-    pass
+    match_list = []
+
+    field_list_namespace = {
+        field.GetNamespaceName() for field in field_list
+    }
+    assert len(field_list_namespace) == 1, 'Fields not under one namespace'
+    namespace = field_list_namespace.pop()
```

Comment:
this will fail under normal circumstances because nearly all fields are defined in the global namespace (''), but very few types are defined in that namespace.  

I think this matching needs some rethinking.  https://github.com/google/digitalbuildings/blob/617eb8ab4739a0c8dad89ca81d1c210dbd221192/tools/validators/ontology_validator/yamlformat/validator/entity_type_manager.py#L86  implements a completely general greatest-common-subset matching algorithm on the ontology.  Maybe you could adapt this approach for your use case?

Reply:
Is this true? I was under the impression that the namespace validator assigned fields to namespaces during field expansion.